### ISSUE TITLE
Reveal.js with Julia: escape codes are not handled in output of executable code blocks  

### ISSUE THREAD

**Issue by mroavi** [quarto-dev/quarto-cli#929]
Here is a minimal working example to reproduce the problem.
Note how the escape codes are not being handled but instead printed literally.

    ---
    format:
      revealjs:
        code-fold: false
        echo: true
    jupyter: julia-1.7
    ---
    
    ```{julia}
    quote
      a = b + c
      d = e - f
    end
    ```

![image](https://user-images.githubusercontent.com/19517248/169253113-e5cd514e-5686-4ffc-a939-2b7744ab3455.png)

This is what the output should look like:

![image](https://user-images.githubusercontent.com/19517248/169253278-6edd8dec-5076-498a-8b9b-f6e5d3d27ffb.png)

Here is another example that uses the BenchmarkTools package to benchmark a function call:

    ---
    format:
      revealjs:
        code-fold: false
        echo: true
    jupyter: julia-1.7
    ---
    
    ```{julia}
    using BenchmarkTools
    @benchmark rand(100,100)
    ```

![image](https://user-images.githubusercontent.com/19517248/169254030-60852f30-169f-47cd-946e-bb780b16d01a.png)

What the output should look like:

![image](https://user-images.githubusercontent.com/19517248/169254368-b1b4c34d-1d85-4316-a3b2-1cd6d9d02f78.png)

**Comment by cscheid** [comment]
We will add support for color in the future (so I'm keeping this open) but for now we at least strip the ansi escape codes, so the output is more readable.

**Comment by mroavi** [comment]
ANSI escapes were stripped from plain text output here: https://github.com/quarto-dev/quarto-cli/commit/cd62560b2db5ed7c5f6a2994d19fb5bddcc9e79d

This is how it looks now:

![image](https://user-images.githubusercontent.com/19517248/169362364-4c3b58de-69bf-401d-8af3-03b4f9652c43.png)

Ignoring the colors, I still see some discrepancies between this image and the one produced by Jupyeter. The shape of the characters looks different. Does this have to do with Unicode support?

**Comment by cscheid** [comment]
Yes, that's a font issue. Matching the graphical characters exactly to jupyter's particular output in this case would mean we can't support our own styling requirements. 

But if you seek precise matching of those characters, you can override quarto's CSS with your own for code-formatted output.

**Comment by cscheid** [comment]
As of 2bd7b8719ecf8916f5754c0f3ac63ba5d7597c31, text output from jupyter engines containing escape codes now renders with `ansi-*` classes, when in HTML output. By default we use the same stylesheet as IPython. But if you want to override that behavior, we wrap the `pre` element in a div with class `ansi-escaped-output`. This gives you a high-specificity selector to target: `div.ansi-escaped-output pre` for global changes, and `div.ansi-escaped-output pre span.ansi-*` for the particular colors and styles. This is the output I see now:

<img width="847" alt="image" src="https://user-images.githubusercontent.com/285675/169385605-1d0bbffb-d534-499b-8a63-2efa2653e57a.png">

For non-HTML output, we presently just strip the escape codes. But it's on our radar to fix this for all formats.

**Comment by mroavi** [comment]
Thanks a lot for fixing this so quickly! It is looking pretty good right now.

**Comment by jjallaire** [comment]
@gaborcsardi We just implemented support for ansi color codes in quarto textual output. We currently have this working for Jupyter (see above). It seems like we should also be able to get this working for knitr as well w/ some combination of quarto and crayon being aware of each other and cooperating (e.g. via knit_print methods or hooks that we can install in quarto). LMK what you think.

**Comment by gaborcsardi** [comment]
Yeah, definitely. Right now crayon/cli rules out colors early in the detection process:
https://github.com/r-lib/cli/blob/072ed21823cdedb5eb63adcc9dbd2685c8252383/R/num-ansi-colors.R#L77

You can force colors with an option or env var: https://github.com/r-lib/cli/blob/072ed21823cdedb5eb63adcc9dbd2685c8252383/R/num-ansi-colors.R#L39-L47
The problem with forcing is embedding, if you start another evaluation context or process, it might inherit these settings.

So often it is better to have crayon/cli auto-detect the configuration that supports colors, so e.g. if quarto is running and ANSI sequences are supported from a certain quarto version, we could detect that.

**Comment by jjallaire** [comment]
@cscheid Let's prototype this by setting the option/env var, then if it works well enough we can have crayon/cli check for being rendered under Quarto.

@gaborcsardi Prior to stable release (in June) we are always encouraging/requiring people to update to the latest Quarto, so if we get this in prior to then I don't think we need a version check.

**Comment by gaborcsardi** [comment]
Sounds good. Let me know if you want me to add quarto auto-detection to crayon/cli. That could handle the cases when a subprocess or another knitr run is started from within quarto/knitr.

**Comment by jjallaire** [comment]
Okay, will do once we have the prototype working. In terms of detecting Quarto, you can use these global knitr options: https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/rmd/execute.R#L207-L210 (either `quarto.version >= 1` or `rmarkdown.version >= 3` will detect Quarto)

**Comment by jjallaire** [comment]
Hi there @mroavi, posting back here because I'm giving a talk on Quarto at JuliaCon next month and was hoping to find some good examples of using Julia w/ Quarto. If you are able to share any of your own (or are aware of others) please LMK over on this discussion thread: https://github.com/quarto-dev/quarto-cli/discussions/1261. Thanks!

**Comment by mroavi** [comment]
Hi @jjallaire. Nice! I'm actually planning on giving a talk at JuliaCon too and I will be using Quarto for the slides!

**Comment by mcanouil** [comment]
I believe this issue is no longer present as of Quarto 1.7.13 for Jupyter (this is also mentioned in the thread).

<table>
<tr><th>Input</th><th>Output</th></tr>
<tr><td>

````qmd
---
format:
  revealjs:
    code-fold: false
    echo: true
jupyter: julia-1.11
---

```{julia}
quote
  a = b + c
  d = e - f
end
```

````

</td><td>

<img width="727" alt="Image" src="https://github.com/user-attachments/assets/18390da0-d8b1-4ddc-b1f5-f573669eb86d" />

</td></tr>
</table>

Regarding ANSI support for `knitr` engine, there is a more explicit issue about it, maybe worth closing this one as completed and continues on the one below:

- https://github.com/quarto-dev/quarto-cli/issues/5746

**Comment by cderv** [comment]
Thanks for checking @mcanouil !
